### PR TITLE
enhancement: use a client certificate to establish a TLS connection

### PR DESCRIPTION
**Pull request by kgbu**
## configuration parameter added

- broker
    - client_cert : path of the certificate of client (signed by the CA which broker relies on)
    - client_key : path of the private key to generate client certificate

## testcase added

- TestTLSClientCertConnectLocalPub : just publish message via TLS connection
- TestTLSClientCertConnectLocalPubSub : subscribe via TLS connection to receive published message

## will test case driver fixed. 

added testcases revealed hidden bug of the code.

### CODE REVIEW

File: broker/broker.go [hunk 1]
```diff
@@ -89,22 +91,31 @@ func NewTLSConfig(b *Broker) (*tls.Config, error) {
...
+	ret := &tls.Config{
 		RootCAs:    certPool,
 		ClientAuth: tls.NoClientCert,
 		ClientCAs:  nil,
 		// InsecureSkipVerify = verify that cert contents
 		// match server. IP matches what is in cert etc.
 		InsecureSkipVerify: true,
-	}, nil
+	}
+	if b.ClientCert != "" {
```
Comment (shirou) [inline broker/broker.go]: I think b.ClientKey also should be checked.

File: tests/will_test.go [hunk 2]
```diff
@@ -108,7 +108,7 @@ func TestNoWillSubscribePublishClose(t *testing.T) {
 	    payload = "Hello will just publish world."
 `
 	expectedWill := false
-	ok := genericWillTestDriver(t, configStr, "testnowillafterclose/will", []byte(""), expectedWill)
+	ok := genericWillTestDriver(t, configStr, "/testnowillafterclose/will", []byte(""), expectedWill)
```
Comment (shirou) [inline tests/will_test.go]: Could you tell me why `/` is added at the head of the topic?

File: broker/broker.go (hunk 1 above)
Comment (kgbu) [inline broker/broker.go]: You are right.
I'll check presence of  b.ClinetKey before calling tls.LoadX509KeyPair()

**Comment by shirou** [comment]
`genericWillTestDriver` should be wait 

1. willMsg received  or
2. timeout received

not to sleep fixed time.Second at https://github.com/shiguredo/fuji/blob/develop/tests/will_test.go#L277. 

And this function always retuned false. But at tests, checking the return value. So, this makes tests disabled. Could you change to use channel and return false if the test failed.

File: tests/will_test.go (hunk 2 above)
Comment (kgbu) [inline tests/will_test.go]: In Fuji, the topic string of will message begins with '/' if no topic_prefix is set.

Older testcase sets wrong topic string. 
This has no-effect if will_message is not set.

But previous code set will flag even if will_message is not set.
This fix is to detect such a case.

**Comment by kgbu** [comment]
I'll try to fix genericWillTestDriver to wait willMessage and timeout without goroutine.

File: tests/will_test.go [hunk 3]
```diff
@@ -234,30 +239,48 @@ func genericWillTestDriver(t *testing.T, configStr string, expectedTopic string,
...
+	}
+	f.Sync()
+
+	// execute fuji as external process
+	fujiPath, err := filepath.Abs("../fuji")
+	if err != nil {
+		t.Error("file path not found")
+	}
+	fmt.Println(fujiPath)
+	fmt.Println(tmpTomlName)
```
Comment (shirou) [inline tests/will_test.go]: Are these lines just for debugging? If so, could you delete these?

Comment (kgbu) [inline tests/will_test.go]: sure. I will remove `fmt.Println` lines.

File: tests/will_test.go [hunk 4]
```diff
@@ -271,11 +294,22 @@ func genericWillTestDriver(t *testing.T, configStr string, expectedTopic string,
...
+
+	// wait for startup of external command process
+	time.Sleep(time.Second * 1)
+
+	// kill publisher
+	err = cmd.Process.Kill()
+	if err != nil {
+		t.Error(err)
+	}
+	fmt.Println("broker killed for getting will message")
```
Comment (shirou) [inline tests/will_test.go]: This printf is useful, but `T.Log` (https://golang.org/pkg/testing/#T.Log) is better.

Comment (kgbu) [inline tests/will_test.go]: Thanks for your suggestion. Absolutely true.

Comment (kgbu) [inline tests/will_test.go]: modified to use T.Log.  After this change, tests run without any error.

**Comment by shirou** [comment]
Thank you! LGTM.

**Pull request merged**